android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/RibCoroutineWorker.kt

```diff
@@ -0,0 +1,187 @@
+      // Cancel unbindJob if `bindJob` has completed exceptionally. This is important to abort `onStart` if `bindJob`
+      // gets cancelled externally.
+      // Notice that in most cases, `unbindJob` will already fail when `bindJob` fails (`onStart` throws) by means of
+      // structured concurrency, but that does not happen on normal cancellation.
+      // This `bindJob` cancellation upon the `unbindJob` cancellation is also already set through structured concurrency.
+      it.invokeOnCompletion { throwable ->
+        if (throwable != null) cancel("Bind job has completed exceptionally", throwable)
+      }
+    }
+    // Verify cancellation and dispatch (if needed) using dispatcher in context instead of continuing on current thread.
+    dispatchIfNeeded()
+    bindAndAwaitCancellation(this, worker, bindJob!!)
+  }
+  // We know deferred is non-null because binding coroutine was started undispatched.
+  return BindWorkerHandleImpl(bindJob!!, unbindJob)
+}
+
+private suspend inline fun dispatchIfNeeded() {
+  suspendCoroutineUninterceptedOrReturn sc@{ cont ->
+    val context = cont.context
+    val dispatcher = context[ContinuationInterceptor] as CoroutineDispatcher
+    if (!dispatcher.isDispatchNeeded(context)) return@sc Unit
+    // Coroutine was not in the right context -- we'll dispatch.
+    context.ensureActive()
+    cont.intercepted().resume(Unit)
+    COROUTINE_SUSPENDED
+  }
+  coroutineContext.ensureActive() // Don't continue if coroutine was cancelled after returning from dispatch.
+}
+
+@Suppress("TooGenericExceptionCaught") // Exception is not swallowed
+private suspend fun bindAndAwaitCancellation(
+  scope: CoroutineScope,
+  worker: RibCoroutineWorker,
+  bindJob: CompletableJob,
+) {
+  val childScope = scope + SupervisorJob(scope.coroutineContext.job)
+  try {
+    worker.onStart(childScope)
+    coroutineContext.ensureActive() // Before completing the binding deferred, make sure the binding coroutine is active.
+    bindJob.complete()
+    awaitCancellation() // Never returns normally, so we are sure an exception will be caught.
+  } catch (t: Throwable) {
+    bindJob.cancelOrCompleteExceptionally(t)
+    worker.onStop(t)
+  }
+}
+
+/** Cancel the deferred if [throwable] is a [CancellationException], otherwise completes it exceptionally. */
+private fun CompletableJob.cancelOrCompleteExceptionally(throwable: Throwable) {
+  when (throwable) {
+    is CancellationException -> cancel(throwable)
+    else -> completeExceptionally(throwable)
+  }
+}
+
+private class BindWorkerHandleImpl(
+  bindJob: Job,
+  private val unbindJob: Job,
+) : BindWorkerHandle, Job by bindJob {
+  override fun unbind(): Job {
+    unbindJob.cancel("Worker was manually unbound.")
+    return unbindJob
+  }
+}
+
+// ---- RibCoroutineWorker <-> Worker adapters ---- //
+
+/** Converts a [Worker] to a [RibCoroutineWorker]. */
+public fun Worker.asRibCoroutineWorker(): RibCoroutineWorker = WorkerToRibCoroutineWorkerAdapter(this)
```

Comment:
sorry missed this earlier. Wondering if we should be adding this now given that old presidio Worker now defines coroutineDispatcher too (the resulting threading type may be confusing to the user).Also at this moment we would allow binding existing worker via RibCoroutine binder without a monitoring option

Reply:
> Wondering if we should be adding this now given that old presidio Worker now defines coroutineDispatcher too

I think yes, because we will still have (internal) components that expect a `Worker` in the signature so in the short term we can expect some `Worker` <-> `RibCoroutineWorker` conversions to happen. It's an easy way to plug-in the new contract into existing components expecting the old contract without requiring refactoring.

> Also at this moment we would allow binding existing worker via RibCoroutine binder without a monitoring option

Yes, but it's a general problem with coroutines we need to tackle. It's not as trivial as adapting the current implementation for `Worker` since monitoring `suspend fun` is not as simple as `measureTimeMillis` the function call because of the suspension capability (what if a function suspends for a long time? There's no harm in that as it's not occupying thread resources).

I'm planning on kicking off a general coroutine monitoring proposal soon, to be pluggable into our dispatchers so we can monitor not only workers, but basically any coroutine running on the system.

Reply:
Got it, thanks a lot for the clarification. It make senses to me.

The only thing that could be a bit unclear and didn't add to the previous WorkerBinder PR was the capability to specify the coroutineDispatcher at WorkerBinder.bind methods given it could bring some confusion to the resulting dispatcher at Worker. e.g. WorkerBinder.bind with `RibDispatchers.IO` and `Worker` coroutineDispatcher override with `RibDispatchers.Main`.This would require some tweaks on the binding logic to always honor the dispatcher defined on the Worker but maybe will be good to bring this back so Worker still has the last preference on defining its dispatcher given binding a list of workers on WorkerBinder.bind call (likely to support pluginPoint)  is a frequent pattern on our monorepo

In the adapter case `Worker.asRibCoroutineWorker(): RibCoroutineWorker`, there could be a `Worker` being built to have an expectation to run on UI thread (overriding default Unconfined explicitly with Main). If the resulting `RibCoroutineWorker` is bound in different dispatcher, it won't honor the dispatcher defined at original Worker

Reply:
Actually, the custom dispatcher in Worker is probably not a bit deal for Worker<>RibCoroutineWorker conversion. 

If there is an intention to convert a Worker to RibCouroutineDispatcher is kind of implicit that now you have better control of the dispatchers being used at binding

Reply:
Got it! I figure we can probably have a corresponding property in the `RibCoroutineWorker` to specify the dispatcher, remove the `context` parameter in `CoroutineScope.bind`, and ignore the dispatcher in `CoroutineScope` (if any). That would make both impls very similar (and easily convertible between themselves without losing semantics) and also I think it's a simpler mental model when there's a single place to decide the dispatcher.

Thanks for the feedback! Will re-ask for review once patched.

Reply:
Hey, revisiting this.

On further thought:

1. It does make sense to provide a `val coroutineDispatcher` in `Worker` but it does not make sense in `RibCoroutineWorker`, because `RibCoroutineWorker` can always override the context with `withContext` (since `onStart` is `suspend`). There's no need for it to be a `val` of the interface. This also solves the issue of binding multiple workers at once: every worker can become the SoT of execution context with `withContext`.
```kotlin
override suspend fun onStart(scope: CoroutineScope) = withContext(RibDispatchers.Main) {
  // ...
}
```

Synchronous `Worker` cannot use `withContext` on `onStart`, so it needs to be a `val` on there.

2. It stills makes sense for specifying a default dispatcher in `CoroutineScope.bind`, because if the worker does not care on which dispatcher it runs, *we want it to run on RibDispatchers.Default by default*.

3. There's a fair point on `Worker.asRibCoroutineWorker()` conversion: `Worker` provides a `val coroutineDispatcher` to specify the context of execution of `onStart`. *We must still respect that after conversion*. So I'm patching the conversor with:
```kotlin
  override suspend fun onStart(scope: CoroutineScope) {
    withContext(worker.coroutineDispatcher) {
      worker.onStart(scope.asWorkerScopeProvider())
    }
  }
```

Reply:
Great point and approach for honoring Worker coroutine dispatcher on the conversions. Looks great to me